extensions/assetLibrary/src/client/Library.tsx

```diff
@@ -0,0 +1,466 @@
+
+  const [selectedItem, setSelectedItem] = useState<LibraryRef>();
+  const [working, setWorking] = useState(false);
+  const [addDialogHidden, setAddDialogHidden] = useState(true);
+  const httpClient = useHttpClient();
+  const API_ROOT = '';
+
+  const programmingLanguages = [
+    {
+      key: 'c#',
+      text: 'C#',
+    },
+    {
+      key: 'js',
+      text: 'Javascript',
+    }
+  ];
+
+  const onChangeLanguage = (ev, op, idx) => {
+    setProgrammingLanguageSelection(op.key);
+    return true;
+  }
+
+  const installComponentAPI = async (projectId: string, packageName: string, version: string, isUpdating: boolean) => {
+    return httpClient.post(`${API_ROOT}/projects/${projectId}/import`, {
+      package: packageName,
+      version: version,
+      isUpdating,
+    });
+  };
+
+  const getLibraryAPI = async () => {
+    return httpClient.get(`${API_ROOT}/library`);
+  };
+
+  const getInstalledComponentsAPI = async (projectId: string) => {
+    return httpClient.get(`${API_ROOT}/projects/${projectId}/installedComponents`);
+  };
+
+  const uninstallComponentAPI = async (projectId: string, packageName: string) => {
+    return httpClient.post(`${API_ROOT}/projects/${projectId}/unimport`, {
+      package: packageName,
+    });
+  };
+
+  const isCompatible = (component) => {
+    return (component.language === programmingLanguageSelection);
+  }
+
+  /**
+   * This method will eventually be moved to the server, when we can store this list on disk via an extension mechanism.
+   * For now, we're doing it in the client, using LocalStorage.
+   * The goal is to capture a list of recently used components, and offer them up across projects
+   * @param componentList
+   */
+  const updateRecentlyUsed = (componentList) => {
+
+    componentList.forEach((component) => {
+      if (!recentlyUsed.find((used) => used.name === component.name)) {
+        recentlyUsed.unshift({...component, language: runtimeLanguage });
+      }
+    });
+
+    window.localStorage.setItem(RECENTLY_USED_KEY, JSON.stringify(recentlyUsed));
+    setRecentlyUsed(recentlyUsed);
+
+  }
+
+  useEffect(() => {
+
+
+    getLibraries();
+    const recent = window.localStorage.getItem(RECENTLY_USED_KEY);
+    if (recent) {
+      try {
+        const list = JSON.parse(recent);
+        setRecentlyUsed(list);
+      } catch(err) {
+        window.localStorage.removeItem(RECENTLY_USED_KEY);
+      }
+    }
+
+    if (settings.runtime && settings.runtime.customRuntime === true && settings.runtime.path) {
+      setEjectedRuntime(true);
+      getInstalledLibraries();
+
+      // detect programming language.
+      // should one day be a dynamic property of the runtime or at least stored in the settings?
+      if (settings.runtime.key === 'node-azurewebapp') {
+        setRuntimeLanguage('js');
+        setProgrammingLanguageSelection('js');
+      } else {
+        setRuntimeLanguage('c#');
+        setProgrammingLanguageSelection('c#');
```

Comment:
These seem duplicative. Can they be combined into a single state value?

Reply:
the runtime language determines what type of packages are installable.  the programming language selection determines what is visible in the browse tab. they can be different...